oracle: accept pre-packed trait membership messages

wchargin opened the pull request:
External callers are now expected to provide trait membership data in
the same bit-packed form used for storage internally. This is cheaper
for two reasons. First, it makes the calldata much smaller: previously,
each token ID took up a full word, of which at least 236 bytes were
zero, costing between 1024 and 1252 gas per token ID. Now, each marginal
token ID costs just one set bit, at a marginal cost of 12 gas. Second,
the contract has less busy work to do: it operates on a whole storage
word at once instead of masking in one bit at a time from the input.

This is a change to the contract interface, but the trait oracle SDK
provides a convenience function to convert a message with explicit token
IDs to a packed message.

This also suggests a natural way to handle efficient finalization:
adjoin to `TraitMembershipWord` a `bool finalized` field that finalizes
a single word at a time. Finalization states can themselves be packed
into bitsets, so each storage word describes the finalization state of
65536 token IDs (and in practice we shouldn't need more than one such
storage word), and a project with mint cap `n` can be considered
effectively finalized if its first `ceiling(n / 256.0)` trait membership
words are finalized.

Gas metrics, per `node scripts/gas.js`:

test case             | before  | after  | saved (gas) | saved (%)
----------------------|--------:|-------:|------------:|---------:
Paddle (12)           | 118817  | 108436 | 10381       | 1.10×
Paddle (again; no-op) | 56299   | 47101  | 9198        | 1.20×
256 consecutive       | 290955  | 49872  | 241083      | 5.83×
256 scattered         | 465322  | 242013 | 223309      | 1.92×

wchargin-branch: oracle-bitpack-input-messages

wchargin committed:
oracle: accept pre-packed trait membership messages

External callers are now expected to provide trait membership data in
the same bit-packed form used for storage internally. This is cheaper
for two reasons. First, it makes the calldata much smaller: previously,
each token ID took up a full word, of which at least 236 bytes were
zero, costing between 1024 and 1252 gas per token ID. Now, each marginal
token ID costs just one set bit, at a marginal cost of 12 gas. Second,
the contract has less busy work to do: it operates on a whole storage
word at once instead of masking in one bit at a time from the input.

This is a change to the contract interface, but the trait oracle SDK
provides a convenience function to convert a message with explicit token
IDs to a packed message.

This also suggests a natural way to handle efficient finalization:
adjoin to `TraitMembershipWord` a `bool finalized` field that finalizes
a single word at a time. Finalization states can themselves be packed
into bitsets, so each storage word describes the finalization state of
65536 token IDs (and in practice we shouldn't need more than one such
storage word), and a project with mint cap `n` can be considered
effectively finalized if its first `ceiling(n / 256.0)` trait membership
words are finalized.

Gas metrics, per `node scripts/gas.js`:

test case             | before  | after  | saved (gas) | saved (%)
----------------------|--------:|-------:|------------:|---------:
Paddle (12)           | 118817  | 108436 | 10381       | 1.10×
Paddle (again; no-op) | 56299   | 47101  | 9198        | 1.20×
256 consecutive       | 290955  | 49872  | 241083      | 5.83×
256 scattered         | 465322  | 242013 | 223309      | 1.92×

wchargin-branch: oracle-bitpack-input-messages
wchargin-source: e7530adf454a6ba66a9ae9a3bb202c6a5677159d

teamdandelion commented on sdk/oracle.js:
```diff
@@ -55,12 +59,37 @@ const sign712 = Object.freeze({
   },
 });
 
+function buildAddTraitMemberships(msg) {
```
changes here, as discussed offline.

wchargin replied:
Done. Thanks for the suggestion.

wchargin committed:
[oracle-bitpack-input-messages: factor word-packing helper more cleanly]

wchargin-branch: oracle-bitpack-input-messages
wchargin-source: 758dc8fe05c5efd0bae171dd2f762770883989e6
